src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/RazorEngine.cs

```diff
@@ -31,13 +32,16 @@ internal RazorEngine(ImmutableArray<IRazorEngineFeature> features, ImmutableArra
         }
     }
 
-    public void Process(RazorCodeDocument codeDocument)
+    public void Process(RazorCodeDocument codeDocument, CancellationToken cancellationToken = default)
     {
         ArgHelper.ThrowIfNull(codeDocument);
 
+        cancellationToken.ThrowIfCancellationRequested();
+
         foreach (var phase in Phases)
         {
-            phase.Execute(codeDocument);
+            phase.Execute(codeDocument, cancellationToken);
```

Comment:
Nit: can you just move `ThrowIfCancellationRequested` above the execute line so as not to call it twice? Presumably if the cancellation happens during the final phase there isn't any need to cancel after it anyway?